smart_open/s3.py

```diff
@@ -644,6 +665,134 @@ def __repr__(self):
         )
 
 
+class SinglepartWriter(io.BufferedIOBase):
+    """Writes bytes to S3 using the single part API.
+
+    Implements the io.BufferedIOBase interface of the standard library.
+
+    This class buffers all of its input in memory until its `close` method is called. Only then will
+    the data be written to S3 and the buffer is released."""
+
+    def __init__(
+            self,
+            bucket,
+            key,
+            session=None,
+            resource_kwargs=None,
+            singlepart_upload_kwargs=None,
+            ):
+
+        self._session = session
+        self._resource_kwargs = resource_kwargs
+
+        if session is None:
+            session = boto3.Session()
+        if resource_kwargs is None:
+            resource_kwargs = {}
+        if singlepart_upload_kwargs is None:
+            singlepart_upload_kwargs = {}
+
+        self._singlepart_upload_kwargs = singlepart_upload_kwargs
+
+        s3 = session.resource('s3', **resource_kwargs)
+        try:
+            self._object = s3.Object(bucket, key)
+            s3.meta.client.head_bucket(Bucket=bucket)
+        except botocore.client.ClientError:
+            raise ValueError('the bucket %r does not exist, or is forbidden for access' % bucket)
+
+        self._buf = io.BytesIO()
+        self._total_bytes = 0
+        self._closed = False
+
+        #
+        # This member is part of the io.BufferedIOBase interface.
+        #
+        self.raw = None
+
+    def flush(self):
+        pass
+
+    #
+    # Override some methods from io.IOBase.
+    #
+    def close(self):
+        self._buf.seek(0)
+
+        try:
+            self._object.put(Body=self._buf, **self._singlepart_upload_kwargs)
+        except botocore.client.ClientError:
+            raise ValueError(
+                'the bucket %r does not exist, or is forbidden for access' % self._object.bucket_name)
+
+        logger.debug("direct upload finished")
+        self._closed = True
+
+    @property
+    def closed(self):
+        return self._closed
+
+    def writable(self):
+        """Return True if the stream supports writing."""
+        return True
+
+    def tell(self):
+        """Return the current stream position."""
+        return self._total_bytes
```

Comment:
Why don't we use `self._buf.tell()` here? That way, we could get rid of the `_total_bytes` attribute entirely.

Reply:
Nevermind, let's keep it for symmetry.